### PR TITLE
Create a workflow to label a PR based on its linked issues

### LINKED ISSUE #75: [Feature] Automatically label PRs with the aspect and goal labels
## Problem

Automatically set the aspect and goal labels on PRs to be the same as their linked issues.

## Alternatives

We can also have a default "needs labels" label that is applied on PRs by default. That way we won't have to find linked issues and just set this label.

## Additional context

Since GitHub API does not offer a way to quickly find the issue linked to a PR, scraping the issue page is an [inelegant but functional solution](https://stackoverflow.com/a/60780499).

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### LINKED ISSUE #175: PRs created with labels via GH CLI fail label check
## Description

It would appear that there are some cases where the label check fails even though a PR is opened with the correct labels. These failures persist across re-runs of the label check GHA and are only resolved by removing the labels and re-adding them in a separate step.

## Reproduction

1. Submit a PR using the GH CLI (`gh pr create`)
2. During the metadata step, add the required labels (e.g. `priority: medium`, `aspect: dx`, `goal: improvement`)
3. Observe that the PR label check fails even though the labels exist on the PR
4. (Optional) Re-run the label check GHA and observe that it fails again

## Expectation

If a PR is submitted with the appropriate labels, the PR label check should pass.

## Additional context

See WordPress/openverse-catalog#372

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes #175 by @AetherUnbound 
Fixes #75 by @dhruvkb 

## Description

This PR sets up the action to label PRs based on info from their linked issues. It follows this algorithm:

- if the PR has linked issues, iterate over them to find the issue with all 3 categories of labels (`aspect: *`, `goal: *` and `priority: *`)
  - if a fully-labelled linked issue is found, apply the labels from that issue on the PR
- else, apply the `🚦 status: awaiting triage` label.

## Testing Instructions

GitHub actions are notoriously hard to test. But the Python script can be tested locally.

0. Populate the `.env` file with a GH access token in the `ACCESS_TOKEN` variable.

1. Run the script as 
```
pipenv run python label_pr.py --pr-url <any PR url that ACCESS_TOKEN can access>
```

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: python/label_pr.py
Comment (AetherUnbound) [inline python/label_pr.py:139]: Clever!

File: python/label_pr.py
Comment (AetherUnbound) [inline python/label_pr.py:149]: I _think_ this could be replaced with a [`for...else` loop](https://book.pythontips.com/en/latest/for_-_else.html) (we may need to default `linked_issues` to an empty list)

```suggestion
    for issue in linked_issues:
        issue = get_issue(gh, issue)
        labels = issue.labels
        if all(
            [
                (aspect := get_label_of_cat("aspect", labels)),
                (priority := get_label_of_cat("priority", labels)),
                (goal := get_label_of_cat("goal", labels)),
            ]
        ):
            log.info(f"Aspect label: {aspect}")
            log.info(f"Priority label: {priority}")
            log.info(f"Goal label: {goal}")
            pr.set_labels(aspect, priority, goal)
            break
    else:
        log.info("Could not find properly labelled issue")
        pr.set_labels("🚦 status: awaiting triage")
```

**Review by AetherUnbound — COMMENTED** [review]
I like it! So with this new flow, I _should not_ add labels to a PR when creating it from the `gh` CLI, is that correct? (since they'll be added automatically).

It also seems like the PR check will pass if we have `awaiting triage` as a label on the PR, is that the outcome we want? I would assume that the label check should fail if that's the case in order to alert the submitter that the labels should be updated (in case a linked issue couldn't be found). 

Frustrating too that the API doesn't have a easy way to get linked issues!

File: .github/workflows/label_pr.yml
Comment (sarayourfriend) [inline .github/workflows/label_pr.yml:7]: Could we use `github.token` for this (or `GITHUB_TOKEN`) instead? Or does that not have enough permissions? I think it should :thinking: Would be nice to get rid of a secret to manage is all.

File: python/label_pr.py
Comment (sarayourfriend) [inline python/label_pr.py:89]: Very obscure selector, this one! I'm surprised it's reliable but it does seem to be based on my own poking around the page. Hopefully they don't change the margins :stuck_out_tongue:

**Review by sarayourfriend — APPROVED** [review]
This is great! I love the solution you picked and impressed by the whole repository dispatch thing, I had no idea that was possible. Glad our contributors won't have to worry about this any more.

**Review by zackkrida — APPROVED** [review]
This is so awesome! One question: it's not clear to me when the checks will fail now. It seems like in the case where there _isn't_ a linked PR, and the user doesn't add labels manually, we'd want the 'awaiting triage' label added _and_ for the checks to fail, so that a maintainer could update the PR with correct labels.

**Comment by sarayourfriend** [comment]
> and the user doesn't add labels manually, we'd want the 'awaiting triage' label added and for the checks to fail, so that a maintainer could update the PR with correct labels.

I suppose the problem with that is that new contributors who don't have write access to the repo aren't able to take action on the failing checks, which can be frustrating and confusing.

I wonder if we shouldn't disable checks altogether on fork PRs until a maintainer has removed the awaiting triage label and added the correct labels.

**Comment by zackkrida** [comment]
> I suppose the problem with that is that new contributors who don't have write access to the repo aren't able to take action on the failing checks, which can be frustrating and confusing.

True. As long as we don't end up with any unlabeled PRs getting merged, which we can monitor, this should be okay as-is.

> I wonder if we shouldn't disable checks altogether on fork PRs until a maintainer has removed the awaiting triage label and added the correct labels.

That's a really compelling idea! Could be much less stressful for contributors.

File: .github/workflows/label_pr.yml
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:7]: We should make a separate issue to remove the `ACCESS_TOKEN` from all the workflows it's used it and use `GITHUB_TOKEN` in all those places.

File: python/label_pr.py
Comment (dhruvkb) [reply python/label_pr.py:139]: Thanks!

File: .github/workflows/label_pr.yml
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:21]: Nothing specific, I've just not tested the code to work on 3.10 as I'm using 3.9 locally.

File: .github/workflows/label_pr.yml
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:4]: Yeah, I love that this is possible but there's also a lot of things missing, notably the ability to callback at the end of a workflow.

File: python/label_pr.py
Comment (dhruvkb) [reply python/label_pr.py:89]: It is indeed very random, I came across this hack on a forum thread and was very skeptical that a margin selector would be specific enough (but it works perfectly so 🤷 ).

File: python/label_pr.py
Comment (AetherUnbound) [reply python/label_pr.py:78]: Ah good point, I've done that a few times myself

File: .github/workflows/label_pr.yml
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:7]: Created #182.

**Comment by dhruvkb** [comment]
So the idea is that the checks should only fail when someone has the ability to fix them.
- contributors cannot label their PRs so those checks failing for them doesn't make sense
- maintainers can alter labels so failing checks for them has value

So PRs will always be labelled automatically on creation (yay!), either correctly (if linked to an issue) or as '🚦 status: awaiting triage'. For a contributor who cannot edit labels, checks will pass in either case.

As a maintainer, you can edit the labels if the default values are not right according to you, by changing the labels, you can end up in an invalid state (incomplete state but not awaiting triage) so the checks will fail for you.

File: python/label_pr.py
Comment (dhruvkb) [reply python/label_pr.py]: Great use of `for`-`else`. Thanks!

File: .github/workflows/label_pr.yml
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:21]: Created #183 to tackle this.

**Comment by dhruvkb** [comment]
I'm merging this to get the repository dispatch into production, let's not merge the sync PRs right away, as there might be changes needed.

File: python/label_pr.py
Comment (sarayourfriend) [reply python/label_pr.py:78]: Ah, did not consider that!

**Comment by sarayourfriend** [comment]
> True. As long as we don't end up with any unlabeled PRs getting merged, which we can monitor, this should be okay as-is.

To be fair @zackkrida none of our CI workflows are required so the situation there has always been precarious and up to the responsibility of the people merging PRs.

**Pull request merged**
